<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цели-исследование" data-toc-modified-id="Цели-исследование-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цели исследование</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследование таблиц</a></span><ul class="toc-item"><li><span><a href="#Сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Сколько-книг-вышло-после-1-января-2000-года-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-среднюю-оценку-для-каждой-книги" data-toc-modified-id="Количество-обзоров-и-среднюю-оценку-для-каждой-книги-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Количество обзоров и среднюю оценку для каждой книги</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Cреднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Cреднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Анализ сервиса для чтения книг по подписке

## Цели исследование

Проанализировать базу данных, в которой содержится инфрмация о большом количестве книг из сервиса для чтения книг по подписке.

Наша задача - проанализировать базу данных, ответив на следующие вопросы:
 - сколько книг вышло после 1 января 2000 года;
 - количество обзоров и среднюю оценку для каждой книги;
 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок;
 - определим издательство, которое выпустило наибольшее число книг;
 - определим автора с самой высокой средней оценкой книг

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Загрузка данных


In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': '***', # имя пользователя
'pwd': '***', # пароль
'host': '***',
'port': ****, # порт подключения
'db': '***'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Исследование таблиц

In [3]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

In [4]:
query_books = ''' SELECT * 
        FROM books t limit 5
        '''

In [5]:
select(query_books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
query_publishers= ''' SELECT * 
        FROM publishers limit 5
        '''

In [7]:
select(query_publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
query_ratings = ''' SELECT * 
        FROM ratings limit 5
        '''

In [9]:
select(query_ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
query_reviews = ''' SELECT * 
        FROM reviews limit 5
        '''

In [11]:
select(query_reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Сколько книг вышло после 1 января 2000 года

In [12]:
select(query_books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [13]:
query_books = '''
SELECT Count(*) AS number_of_books
FROM   books t
WHERE  t.publication_date > '2000-01-01'
'''

In [14]:
select(query_books)

number_of_books
0              819

**Вывод**

После 1 января 2000 года было выпущено 819 книг, то есть большая часть книг нашей базы данных

### Количество обзоров и среднюю оценку для каждой книги

In [15]:
rat_rev = ''' 
WITH rev
     AS (SELECT r.book_id,
                Count(r.text) AS review_cnt
         FROM   reviews r
         GROUP  BY r.book_id
         ORDER  BY review_cnt DESC),
     rat
     AS (SELECT ra.book_id,
                Count(ra.rating) AS rate_cnt,
                Avg(ra.rating)   AS rate_avg
         FROM   ratings ra
         GROUP  BY ra.book_id)
SELECT rev.book_id,
       b.title,
       review_cnt,
       rate_cnt,
       rate_avg
FROM   rat
       LEFT OUTER JOIN rev
                    ON rat.book_id = rev.book_id
       LEFT JOIN books b
              ON rat.book_id = b.book_id
ORDER  BY rate_cnt DESC 
                    
'''

In [16]:
select(rat_rev)

book_id                                              title  review_cnt  \
0      948.0                            Twilight (Twilight  #1)         7.0   
1      750.0                The Hobbit  or There and Back Again         6.0   
2      673.0                             The Catcher in the Rye         6.0   
3       75.0               Angels & Demons (Robert Langdon  #1)         5.0   
4      302.0  Harry Potter and the Prisoner of Azkaban (Harr...         6.0   
..       ...                                                ...         ...   
995    175.0                                  Dance Dance Dance         2.0   
996    403.0  Lord John and the Private Matter (Lord John Gr...         2.0   
997    400.0                   London Bridges (Alex Cross  #10)         2.0   
998    401.0                      Long Day's Journey into Night         2.0   
999      NaN  Disney's Beauty and the Beast (A Little Golden...         NaN   

     rate_cnt  rate_avg  
0         160  3.662500  
1          88  4.125000  
2          86  3.825581  
3          84  3.678571  
4          82  4.414634  
..        ...       ...  
995         2  4.500000  
996         2  3.500000  
997         2  4.000000  
998         2  4.000000  
999         1  4.000000  

[1000 rows x 5 columns]

**Вывод**

Не у каждой книги есть обзор, но оценки есть у каждой. Первые "Сумерки" лидируют по количеству написанных обзоров, однако оценили её всего на 3.66

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [17]:
pubs = '''
SELECT p.publisher,
       Count(book_id) AS number_of_books
FROM   books t
       LEFT JOIN publishers p
              ON t.publisher_id = p.publisher_id
WHERE  t.num_pages > 50
GROUP  BY p.publisher
ORDER  BY number_of_books DESC
LIMIT  1 
'''

In [18]:
select(pubs)

publisher  number_of_books
0  Penguin Books               42

**Вывод**

Из издательств, представленных в базе данных, больше всех книг выпустило издательство "Penguin Books" - 42 публикации больше 50 страниц

### Автор с самой высокой средней оценкой книг

In [19]:
author = '''
WITH  br AS
          (SELECT    t.book_id,
                    Count(r.rating) AS rate_cnt,
                    Avg(r.rating)   AS rating_avg
          FROM      books t
          LEFT JOIN ratings r
          ON        t.book_id = r.book_id
          GROUP BY  t.book_id
          HAVING    Count(r.rating) > 50)
SELECT    a.author,
          Avg(br.rating_avg) AS book_rate_avg
FROM      br
LEFT JOIN books b
ON        br.book_id = b.book_id
LEFT JOIN authors a
ON        b.author_id = a.author_id
GROUP BY  a.author
ORDER BY  book_rate_avg DESC limit 1
'''

In [20]:
select(author)

author  book_rate_avg
0  J.K. Rowling/Mary GrandPré       4.283844

**Вывод**

Автором с самой высокой средней оценкой книги является J.K. Rowling/Mary GrandPré, автор всеми любимых книжек про мальчика - волшебника

### Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [21]:
users = '''
WITH c
     AS (SELECT r.username,
                Count(r.rating) AS rate_cnt
         FROM   ratings r
         GROUP  BY r.username
         HAVING Count(r.rating) > 50),
     h
     AS (SELECT re.username,
                Count(re.review_id) AS review_cnt
         FROM   reviews re
         GROUP  BY re.username)
SELECT Avg(h.review_cnt) AS review_cnt_avg
FROM   h
       INNER JOIN c
               ON h.username = c.username '''

In [22]:
select(users)

review_cnt_avg
0       24.333333

**Вывод**

Среднее количество обзоров от тех пользователей, которые поставили более 50 оценок, равно 24.3333. Люди, которые много читают, любят писать ревью на прочитанное

## Общий вывод

 - большая часть книг была выпущена в последние 22 года, то есть в основном в сервисе представлены авторы современности
 - для каждой книги мы посчитали количество обзоров и среднюю оценку
 - определили, что penguine books выпустило 42 книги, это составляет 4% от всех книг в базе данных
 - автором с самой высокой средней оценкой (с количеством отзывов более 50) является Роулинг
 - люди, которые много читают, любят писать ревью на прочитанное, мы видим, что у пользователей с оценками от 50 количество отзывов в среднем 24 штуки.

Дальнейший анализ представленной базы данных поможет решить следующие большие вопросы нового продукта:

 - Помочь рекламе - для привлечения пользователей мы можем составить их портрет, в рекламе предлагать картинки из произведений, которые им нравятся
 - Помочь дизайну - у нас есть книги, которые можно поместить на главный экран - любимые и самые обсуждаемые тайтлы
 - Удержать пользователя - тем пользователям, которые уже читают нас можем предложить другие книги этого автора или этого издательства, чтобы они дальше продлевали подписку.